### PR TITLE
Build the Olm bundle and test it in CI

### LINKED ISSUE #10908: Prepare scripts for generating OLM resources
This includes tests in GH Actions as well

### PR DESCRIPTION
**Pull request by andreaTP**
Resolves #10908

### CODE REVIEW

**Comment by andreaTP** [comment]
@stianst before merging this job we need those secrets:

```bash
TEST_DOCKER_REGISTRY=quay.io
TEST_DOCKER_USERNAME=keycloak-testing #?
TEST_DOCKER_TOKEN=<you-know-what-to-do>
```

File: operator/olm-base/metadata/annotations.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+annotations:
+  # Core bundle annotations.
+  operators.operatorframework.io.bundle.mediatype.v1: registry+v1
+  operators.operatorframework.io.bundle.manifests.v1: manifests/
+  operators.operatorframework.io.bundle.metadata.v1: metadata/
+  operators.operatorframework.io.bundle.package.v1: keycloak-operator
+  operators.operatorframework.io.bundle.channels.v1: alpha
+  operators.operatorframework.io.bundle.channel.default.v1: alpha
```
Comment (vmuzikar) [inline operator/olm-base/metadata/annotations.yaml:8]: jFYI – no need to change it in this PR. The channel name will have to be different. Probably `candidate` for releases and `nightly` for nightlies. I.e. will have to be variable. We can do it as a follow-up.

File: operator/scripts/prepare-olm-test.sh [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+PREV_VERSION="$3"
+
+OPERATOR_IMAGE_NAME="keycloak-operator"
+OPERATOR_DOCKER_IMAGE="$DOCKER_REGISTRY/$OPERATOR_IMAGE_NAME"
+
+# Create OLM bundle
+$SCRIPT_DIR/create-olm-bundle.sh $VERSION $PREV_VERSION $OPERATOR_DOCKER_IMAGE
+
+# WARNING: This should be done ONLY for tests
+$SCRIPT_DIR/inject-custom-image.sh $VERSION "$DOCKER_REGISTRY/keycloak:$VERSION"
```
Comment (vmuzikar) [inline operator/scripts/prepare-olm-test.sh:18]: No need for this. Env vars can be directly specified in the `Subscription`, see https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/design/subscription-config.md.

File: operator/scripts/inject-custom-image.sh [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#! /bin/bash
+set -euxo pipefail
+
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+
+VERSION="$1"
+VALUE="$2"
+
+yq ea -i ".spec.install.spec.deployments[0].spec.template.spec.containers[0].env += {\"name\": \"OPERATOR_KEYCLOAK_IMAGE\", \"value\": \"$VALUE\"}" $SCRIPT_DIR/../olm/$VERSION/manifests/keycloak-operator.v$VERSION.clusterserviceversion.yaml
```
Comment (vmuzikar) [inline operator/scripts/inject-custom-image.sh:9]: No need for this.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@andreaTP Overall nice work, thanks! Added a few comments. Main blocker from my PoV is the cluster insulation from previous tests and missing "test" for realm import CRD.

File: .github/workflows/operator-ci.yml [hunk 4]
```diff
@@ -65,8 +65,79 @@ jobs:
...
+      #         -Dquarkus.jib.jvm-arguments="-Djava.util.logging.manager=org.jboss.logmanager.LogManager","-Doperator.keycloak.image=keycloak:${GITHUB_SHA}","-Doperator.keycloak.image-pull-policy=Never" \
+      #         -DskipTests
+      # - name: Install the operator CRDs
+      #   working-directory: operator
+      #   run: |
+      #     kubectl apply -f target/kubernetes/keycloaks.keycloak.org-v1.yml
+      #     kubectl apply -f target/kubernetes/keycloakrealmimports.keycloak.org-v1.yml
+
+      # Test the OLM bundle
+      - name: Install OPM
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply .github/workflows/operator-ci.yml:93]: ok in separating it to another job

File: .github/workflows/operator-ci.yml [hunk 5]
```diff
@@ -65,8 +65,79 @@ jobs:
...
+        working-directory: operator
+        run: |
+          kubectl apply -f src/main/resources/example-postgres.yaml
+          kubectl apply -f src/main/resources/example-keycloak.yml
+          # Wait for the deployment to be created
+          timeout 120 sh -c "while ! kubectl get deployment/example-kc; do (sleep 1 && kubectl get pods); done"
+          # Wait for the deployment to be available
+          kubectl wait --for=condition=available --timeout=600s deployment/example-kc
+          # Wait for the pods in the namespace to be ready
+          kubectl wait --for=condition=Ready --timeout=600s pods -l app=keycloak
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply .github/workflows/operator-ci.yml:143]: ok

File: .github/workflows/operator-ci.yml [hunk 6]
```diff
@@ -65,8 +65,79 @@ jobs:
...
+
+      - name: Install the operator with OLM
+        working-directory: operator
+        run: ./scripts/install-keycloak-operator.sh
+
+      - name: Deploy an example Keycloak and wait for it to be ready
+        working-directory: operator
+        run: |
+          kubectl apply -f src/main/resources/example-postgres.yaml
+          kubectl apply -f src/main/resources/example-keycloak.yml
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply .github/workflows/operator-ci.yml:137]: ok, done

File: operator/scripts/inject-custom-image.sh (hunk 3 above)
Comment (andreaTP) [reply operator/scripts/inject-custom-image.sh:9]: if the configuration through the subscription works as expect I'll remove the file entirely

File: .github/workflows/operator-ci.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply .github/workflows/operator-ci.yml:143]: moved to a shell script as well

File: .github/workflows/operator-ci.yml [hunk 7]
```diff
@@ -65,8 +65,79 @@ jobs:
...
+      #     kubectl apply -f target/kubernetes/keycloaks.keycloak.org-v1.yml
+      #     kubectl apply -f target/kubernetes/keycloakrealmimports.keycloak.org-v1.yml
+
+      # Test the OLM bundle
+      - name: Install OPM
+        uses: redhat-actions/openshift-tools-installer@v1
+        with:
+          source: "github"
+          opm: "1.21.0"
+          operator-sdk: "1.18.1"
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply .github/workflows/operator-ci.yml:98]: removed

File: operator/scripts/inject-custom-image.sh (hunk 3 above)
Comment (andreaTP) [reply operator/scripts/inject-custom-image.sh]: removed

File: .github/workflows/operator-ci.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply .github/workflows/operator-ci.yml:171]: don split in 3 separate jobs

File: .github/workflows/operator-ci.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply .github/workflows/operator-ci.yml]: now those checks are performed in the relevant script

File: operator/scripts/prepare-olm-test.sh (hunk 2 above)
Comment (andreaTP) [reply operator/scripts/prepare-olm-test.sh]: works, using `subscription.config` and removing the script

**Comment by andreaTP** [comment]
ready for next round @vmuzikar

File: .github/workflows/operator-ci.yml [hunk 8]
```diff
@@ -48,12 +66,10 @@ jobs:
           github token: ${{ secrets.GITHUB_TOKEN }}
           driver: docker
           start args: '--addons=ingress'
-
       - name: Build the Keycloak Docker image
```
Comment (vmuzikar) [inline .github/workflows/operator-ci.yml:69]: I noticed that we're basically building KC image three times now. Maybe we could build it and push it to Quay right away in the first build step. WDYT?

File: .github/workflows/operator-ci.yml [hunk 9]
```diff
@@ -66,13 +82,143 @@ jobs:
...
+          mvn clean package \
+              -Dquarkus.container-image.build=true \
+              -Dquarkus.container-image.labels."\"quay.expires-after\""="20h" \
+              -DskipTests
+
+      - name: Install the operator CRDs
+        working-directory: operator
+        run: |
+          kubectl apply -f target/kubernetes/keycloaks.keycloak.org-v1.yml
+          kubectl apply -f target/kubernetes/keycloakrealmimports.keycloak.org-v1.yml
```
Comment (vmuzikar) [inline .github/workflows/operator-ci.yml:175]: Why are we installing CRDs? It needs to be done by OLM. If not, then we're doing something wrong. 🙂

**Review by vmuzikar — CHANGES_REQUESTED** [review]

File: .github/workflows/operator-ci.yml (hunk 8 above)
Comment (andreaTP) [reply .github/workflows/operator-ci.yml:69]: it's an option, but building the image is actually pretty fast (~40 seconds) and in this way we do not depend on network on 2/3 of the jobs ...
is a tradeoff, no strong opinion here.

Comment (vmuzikar) [reply .github/workflows/operator-ci.yml:69]: We're uploading the image in any case for the OLM tests, so IMO we can benefit from that in other tests too.

File: .github/workflows/operator-ci.yml [hunk 10]
```diff
@@ -17,29 +17,47 @@ concurrency:
...
     if: ${{ ( github.event_name != 'schedule' ) || ( github.event_name == 'schedule' && github.repository == 'keycloak/keycloak' ) }}
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v2
       - name: Update maven settings
         run: mkdir -p ~/.m2 ; cp .github/settings.xml ~/.m2/
-      - uses: actions/setup-java@v1
+      - uses: actions/setup-java@v2
         with:
+          distribution: 'temurin'
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply .github/workflows/operator-ci.yml:32]: #11017 has been merged, I resolve this one.

File: .github/workflows/operator-ci.yml (hunk 8 above)
Comment (andreaTP) [reply .github/workflows/operator-ci.yml]: done

Comment (andreaTP) [reply .github/workflows/operator-ci.yml]: In the future, we can evaluate if we want to unify even publishing the operator image itself (that can be shared between OLM and remote tests), probably, for this PR, this is enough. wdyt @vmuzikar ?

**Comment by andreaTP** [comment]
@vmuzikar this PR should be ready for a final review, I think I have addressed all the outstanding and this is a link to a green CI:
https://github.com/andreaTP/keycloak/actions/runs/2072358178

**Comment by stianst** [comment]
Quay.io organization created here keycloaktesting, and secrets setup in GitHub

**Comment by andreaTP** [comment]
Thanks a lot @stianst !
I hadn't realized that in Quay.io we need a username that is different from the repository name, I have updated the secrets to be:
```
TEST_DOCKER_REGISTRY=quay.io
TEST_DOCKER_REPOSITORY=keycloaktesting
TEST_DOCKER_USERNAME=<username-to-use-for-login>
TEST_DOCKER_TOKEN=<token-to-login>
```

Would you mind creating the missing secret: `TEST_DOCKER_REPOSITORY=keycloaktesting` ? Thanks in advance!

File: .github/workflows/operator-ci.yml [hunk 11]
```diff
@@ -50,38 +78,121 @@ jobs:
...
+          password: ${{ secrets.TEST_DOCKER_TOKEN }}
+
+      - name: Build and push the operator image
+        working-directory: operator
+        run: |
+          mvn clean package \
+              -Dquarkus.container-image.build=true \
+              -Dquarkus.container-image.push=true \
+              -Dquarkus.container-image.image="${{ secrets.TEST_DOCKER_REGISTRY }}/${{ secrets.TEST_DOCKER_REPOSITORY }}/keycloak-operator:${{ steps.vars.outputs.version }}" \
+              -Dquarkus.container-image.labels."\"quay.expires-after\""="20h" \
```
Comment (andreaTP) [inline .github/workflows/operator-ci.yml:165]: This label is not injected at the moment due to: https://github.com/quarkusio/quarkus/issues/24820

Comment (andreaTP) [inline .github/workflows/operator-ci.yml:165]: This label is not injected at the moment due to: https://github.com/quarkusio/quarkus/issues/24820

File: .github/workflows/operator-ci.yml [hunk 12]
```diff
@@ -50,38 +78,121 @@ jobs:
...
+      - uses: actions/setup-java@v2
+        with:
+          distribution: 'temurin'
+          java-version: ${{ env.JDK_VERSION }}
+          cache: 'maven'
+      - name: Setup Minikube-Kubernetes
+        uses: manusa/actions-setup-minikube@v2.4.3
+        with:
+          minikube version: v1.24.0
+          kubernetes version: v1.22.3
```
Comment (vmuzikar) [inline .github/workflows/operator-ci.yml:110]: Just a nitpick, we can handle this separately. We might want to move the versions to some env var as we install minikube three times now (similarly to `JDK_VERSION`).

**Review by vmuzikar — APPROVED** [review]
@andreaTP LGTM, thanks.

**Review by stianst — APPROVED** [review]

File: .github/workflows/operator-ci.yml (hunk 12 above)
Comment (andreaTP) [reply .github/workflows/operator-ci.yml:110]: that's a good idea, we should refactor a little more next time we work on this file!

**Pull request merged**